### PR TITLE
Decouple forcefield from atomtyping step

### PR DESCRIPTION
**Pull request by umesh-timalsina**
### PR Summary:
This PR introduces `AtomTypingRulesProvider`, which is a data storage python class, holding only the information necessary for atomtyping engine to run. This is one of the many steps necessary to remove the dependecy on `openmm` we currently have.

### CODE REVIEW

File: foyer/atomtyper.py [hunk 1]
```diff
@@ -7,10 +7,37 @@
 from parmed import Structure
 
 from foyer.exceptions import FoyerError
+from foyer.smarts import SMARTS
 from foyer.smarts_graph import SMARTSGraph
 from foyer.topology_graph import TopologyGraph
 
 
+class AtomTypingRulesProvider:
+    """A generic rules provider for atomtyping, agnostic of the forcefield."""
```
Comment (justinGilmer) [inline foyer/atomtyper.py:16]: Can we get a docstring for these init parameters? Other than that, LGTM!

**Review by justinGilmer — CHANGES_REQUESTED** [review]

File: foyer/atomtyper.py (hunk 1 above)
Comment (umesh-timalsina) [reply foyer/atomtyper.py]: Added

File: foyer/atomtyper.py [hunk 2]
```diff
@@ -7,10 +7,50 @@
...
+
+    Parameters
+    ----------
+    atomtype_definitions: dict, required
+        The smarts definition for the atomtypes
+    atomtype_overrides: dict, required
+        The overrides for particular atomtypes
+    non_element_types: set, required
+        The non-element types used for atomtyping
+    parser: The chemical grammer parser, default=None
```
Comment (umesh-timalsina) [inline foyer/atomtyper.py:26]:
```suggestion
    parser: The chemical grammar parser, default=None
```

**Review by daico007 — APPROVED** [review]
LGTM

File: foyer/atomtyper.py [hunk 3]
```diff
@@ -7,10 +7,50 @@
...
+        parser=None,
+    ):
+        self.atomtype_definitions = atomtype_definitions
+        self.atomtype_overrides = atomtype_overrides
+        self.non_element_types = non_element_types
+        self.parser = parser or SMARTS(self.non_element_types)
+
+    @classmethod
+    def from_foyer_forcefield(cls, ff):
+        """Create an instance of the rules provider for an foyer-forcefield."""
```
Comment (justinGilmer) [inline foyer/atomtyper.py:45]:
```suggestion
        """Create an instance of the rules provider for a foyer-forcefield."""
```

**Review by justinGilmer — APPROVED** [review]
One additional change and then LGTM!

**Pull request merged**